Don't use --whole-archive for libc/common and libc/picolibc files

keith-packard opened the pull request:
This moves these two libraries outside of the --whole-archive list so that only functionality explicitly referenced by the application will get included in the executable.

For example -- the malloc implementation in libc/common/source/stdlib/malloc.c includes a `SYS_INIT` function, `malloc_prepare`. If the application references `malloc`, that file will be included in the output, including the `malloc_prepare` function which will get called at startup. If the application does *not* reference `malloc`, then
the file will be excluded and so `malloc_prepare` will not be included in the output.

This is an alternative for #63259 and does not require per-application configuration.

keith-packard committed:
libc/picolibc: Split hooks into separate files

This splits the picolibc helper functions into separate files so that
code not explicitly referenced can be excluded from the result once
we stop using --whole-archive with this library.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/picolibc: Add 'write' implementation when needed

This is needed for the picolibc __chk_fail implementation as that
uses it up through version 1.8.5. After that, picolibc uses puts.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
This might be hard to make work -- having the linker drop code with no references (using --whole-archive --gc-sections) is very different from the traditional model where the linker only adds code with explicit references. It might be worth the effort though -- we're currently pulling in anything marked KEEP in the linker script, even if that's the only bit of a object file which is otherwise unused.

keith-packard commented:
I just added code to compute all of the nested dependencies to stick between --start-group and --end-group; would be happy if someone knew how to use the cmake LINK_GROUP stuff to make that less complicated. Alternatively, figure out how to get cmake to append -Wl,--end-group after all of the dependent libraries.

keith-packard committed:
libc/common: Create __wrap_ aliases for malloc family functions

These aliases will be used to direct libc allocations to the common
allocator during linking using the --wrap linker option.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
ok, another kludge -- I've added z_ symbol aliases for the common malloc functions and then used --defsym to redirect
uses to the common version of malloc instead of the C library version.

keith-packard commented:
and more kludges -- check each link element to see if it's a valid target before attempting to run get_property as that will fail otherwise.

keith-packard commented:
This series only includes patches for the 'ld' linker -- I think it should probably include patches for other linkers as well, but I cannot test all of them.

cfriedt commented on lib/libc/picolibc/assert.c:
```diff
@@ -0,0 +1,45 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <zephyr/posix/sys/stat.h>
+#include <sys/time.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/linker/linker-defs.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/errno_private.h>
+#include <zephyr/sys/libc-hooks.h>
+#include <zephyr/internal/syscall_handler.h>
+#include <zephyr/app_memory/app_memdomain.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/sem.h>
+#include <zephyr/logging/log.h>
+#ifdef CONFIG_MMU
+#include <zephyr/kernel/mm.h>
+#endif
```
Since the same headers are included multiple times by local .c files, would it make sense to put them into a local / private .h file?

The same comment here and elsewhere

keith-packard replied:
alternatively, figure out which headers each file needs. You can probably tell I pulled the original file apart and left every single original header in all of the resulting bits. I'll explore which option looks neater in the end. Thanks for taking a look at this series!

keith-packard replied:
I trimmed down the include lists so that each file only included what it needed. The results look a lot nicer.

cfriedt commented:
Definitely agree with not linking libc with `--whole-archive` 👍

keith-packard commented:
> Definitely agree with not linking libc using `--whole-archive` 👍

This doesn't change that -- we were already placing libc outside of the `--whole-archive` set. This sticks the C library helpers and common bits outside too.

cfriedt commented:
The auto-malloc concept sounds great.

keith-packard commented:
> The auto-malloc concept sounds great.

Let me see if I can't get the allocator aliasing working -- that's required to map malloc calls in the C library back to the common malloc implementation.

keith-packard commented:
> > The auto-malloc concept sounds great.
> 
> Let me see if I can't get the allocator aliasing working -- that's required to map malloc calls in the C library back to the common malloc implementation.

aha! The linker `--wrap` flag is your friend here. That does exactly what I want -- remaps malloc calls to __wrap_malloc *without* pulling malloc in when unused.

cfriedt reviewed (dismissed):
Looks ok to me

keith-packard commented:
> Looks ok to me

Thanks for looking. Any idea what else is needed to get this merged? It seems like a pretty 'normal' way of using libraries.

cfriedt commented:
> > Looks ok to me
> 
> Thanks for looking. Any idea what else is needed to get this merged? It seems like a pretty 'normal' way of using libraries.

I'll drop this link in `#pr-help`

fabiobaltieri commented:
@stephanosio can you take a look

carlescufi commented:
DNM until @tejlmand has taken a look

tejlmand commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -1,7 +1,17 @@
 
 zephyr_library()
-zephyr_library_sources(libc-hooks.c)
+zephyr_library_property(PARTIAL TRUE)
```
it seems wrong to me that we are starting to modify the `zephyr_library()` this way.

`zephyr_library()` is partly a convenience for users, but also to ensure that such libraries are included inside the `--whole-archive` for various reasons.

Libraries that explicitly should not go inside the `--whole-archive` flags should be created as regular CMake libraries, just like it's done for the `kernel` lib.
https://github.com/zephyrproject-rtos/zephyr/blob/998e839d6743c3ae2633e653415a2ae56bb7fb34/kernel/CMakeLists.txt#L96

and then regular CMake linking should be used.

From the PR I don't see the exact issues faced with circular dependencies, but from the proposed solution then i'm assuming that `kernel` uses / references picolibc ( or newlib) which again uses kernel, thus forming a circular dependency.

Such cases is supported natively in CMake, by having the n-static libs linking to each other, like this:
```
target_link_libraries(kernel PUBLIC picolibc) # Kernel depends on picolibc
target_link_libraries(picolibc PUBLIC kernel) # Picolibc depends on kernel
```
thus informing CMake that those two libs should be linking si that circular dependencies are resolved.

Now, if that isn;t sufficient because one link result in new function references to be resolved, then the `INTERFACE_LINK_LIBRARIES` property can be set in addition.
For example: 
```
set_target_properties(kernel PROPERTIES _LINK_INTERFACE_MULTIPLICITY 3) 
set_target_properties(picolibc PROPERTIES _LINK_INTERFACE_MULTIPLICITY 3) 
```

keith-packard replied:
Thanks! I'll give these suggestions a try.

keith-packard replied:
This was going fairly well; I've converted libc/common and libc/picolibc to use add_library and reverted all of the kludges related to zephyr_library usage. However, I've run up against a snag -- I need to include the C library in the repeated library set, so that I get:
```
    zephyr/lib/libc/picolibc/libc_picolibc.a \
    zephyr/lib/libc/common/libc_common.a \
    -lc \
    -lgcc \
    zephyr/kernel/libkernel.a \
    zephyr/lib/libc/picolibc/libc_picolibc.a \
    zephyr/lib/libc/common/libc_common.a \
    -lc \
    -lgcc \
    zephyr/kernel/libkernel.a \
    zephyr/lib/libc/picolibc/libc_picolibc.a \
    zephyr/lib/libc/common/libc_common.a \
```
If I could compute the pathname to libc.a, I'd use that directly. However, that's buried inside the picolibc.specs file and gcc won't print it with `-print-file-name` because that doesn't parse the specs file looking for library paths set within that file.

How can I get cmake to insert `-lc -lgcc` multiple times into the link command line? Alternatively, how can I get cmake to wrap those static libraries in `-Wl,--start-group ... -Wl,--end-group`?

keith-packard replied:
I also tried using `CMAKE_C_LINK_GROUP_USING_RESCAN`, but cmake gets very upset because it generates a circular dependency between kernel and the header generation bits, the latter not being a static library.

keith-packard replied:
Ok, I think this is all generic cmake stuff now. There's a bit of a kludge to make a toolchain C library work correctly, but otherwise it's not terrible. Let me know if this is the direction you're thinking. I'm finding some useful size reductions here.

tejlmand reviewed (changes requested):
thus should be handled using vanilla CMake.

Please link the static libraries which results in circular dependencies against each other, as described in the comment.

An alternative solution can be to discuss bumping the minimum required CMake version, as it has now been some years since we last did this.
If we do so, we can make use of https://cmake.org/cmake/help/latest/variable/CMAKE_LINK_GROUP_USING_FEATURE.html.

Before approving a custom solution to the described problem, then I would like to know if existing CMake features has been tried, and if so, what issues was faced.

keith-packard committed:
kernel: Add kernel library early in the build

Create the kernel library object before loading in the subdirectories so
that other parts of the code (the toolchain and C library) may adjust the
kernel dependencies as needed using target_link_library(kernel, ...)

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc: Switch minimal, common and picolibc to regular libraries

Use add_library instead of zephyr_library and then add a dependency from
the kernel to the C library bits.

This moves all of this code out of the --whole-archive set so that unused
bits from these libraries are not included in the resulting executable,
even sections marked 'KEEP' in the linker script.

Because this changes the names of the C libraries, linker script fragments
in soc/xtensa, soc/riscv and boards/qemu_x86_tiny needed to be updated.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Aha! A terrible kludge and it's working now. The kludge is to create a "fake" C library target and assign it the filename (wait for it) "-lc". Now, simply create a file named `${CMAKE_BINARY_DIR}/-lc` and everyone is happy -- cmake believes the file is ready for use and happily inserts `-lc` in the linker command line as often as necessary.

I did the same thing for -lgcc to make sure any references back to libc work correctly.

keith-packard committed:
west: Pull in picolibc module fixes for library dependencies

Support changes to Zephyr that manage the C library dependencies using
standard cmake mechanisms.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
cmake/linker: Insert zephyr_link_libraries before kernel

As many libraries used in zephyr_link_libraries depend upon C library
and/or kernel interfaces, they need to be present in the link command line
before those libaries.

Do this by adding them to the linker command ahead of the kernel by saving
them in zephyr_link_libraries and using that in toolchain_ld_link_elf.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
And one more (modest) hack -- stick zephyr_link_libraries in front of 'kernel' in the link command line.

keith-packard committed:
kernel/mem_domain: Allow z_libc_partition to be empty

We can't know before building the system whether there will be any data
present in the C library parition. Allow this, simply skipping any use of
it.

Signed-off-by: Keith Packard <keithp@keithp.com>

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -109,6 +109,7 @@ endfunction()
 
 # https://cmake.org/cmake/help/latest/command/target_link_libraries.html
 function(zephyr_link_libraries)
+  set_property(TARGET zephyr_interface APPEND PROPERTY LINK_LIBS ${ARGV})
```
Why do we need to maintain this property ?

Secondly, this will probably break when using generator expressions with `zephyr_link_libraries()` as generator expressions cannot be correctly retrieved on custom properties, only some properties support genex.

If this is the reason:
> As many libraries used in zephyr_link_libraries depend upon C library
and/or kernel interfaces, they need to be present in the link command line
before those libaries.

then what should be done is this:
```
# This is a regular lib depending on Kernel
add_library(some_lib) 

target_link_library(some_lib PUBLIC kernel)
```

that will ensure the following linking order `... some_lib kernel ...` which seems to be what you want to achieve.

@keith-packard can you provide the an example of what is breaking today, cause this proposal could indicate that it's as simple as missing such link dependency.

And I will be happy to assist.

This is one downside of the `zephyr_library()`, that users never get to understand the internals of CMake's linking order because things are whole-archived together, and when they suddenly are not, things break.

keith-packard replied:
As above, there are two problems with trying to use direct dependencies. The first is toolchain libraries, like `-lstdc++` and `-lc` where we cannot use add_library without kludges. The second is that every attempt I made to do this resulted in cmake finding a circular dependency that included `kernel`, `offsets_h`, `offsets`, and `zephyr_interface`.

tejlmand commented on CMakeLists.txt:
```diff
@@ -868,6 +868,13 @@ endforeach()
 get_property(ZEPHYR_LINK_LIBS_PROPERTY TARGET zephyr_interface PROPERTY LINK_LIBS)
 
+# Remove any non-file options which aren't libraries from this property
+foreach(zephyr_link_lib ${ZEPHYR_LINK_LIBS_PROPERTY})
+ if(zephyr_link_lib MATCHES "^-.*" AND NOT zephyr_link_lib MATCHES "^-l.*")
```
and this already gives a hint why this is a dangerous path to be taken.

I believe the general CMake lib linking handling can take care of this for us, but I would need to know the exact issues faced / how I can reproduce, then I can provide a PR proposal adjustment, I believe that will be the most efficient way to tackle this.
Although I'm also busy with HWMv2, then I believe this is worth (and most efficient) to be spending time on.
Also feel free to ping on Discord.

keith-packard replied:
> and this already gives a hint why this is a dangerous path to be taken.

No, this should be replaced by separating this into two functions, one for libraries and the other for link options. I didn't want to get started on that as it's more invasive than this kludge.

keith-packard replied:
I guess the real question is whether this is worth the effort at all. Do we want to get away from `--whole-archive` as the way to resolve the inherent circular dependencies when linking Zephyr? Or do we just want to leave things as they are and expect application developers to configure the system to explicitly exclude bits they don't use?

tejlmand reviewed (changes requested):
I think this can be made much simpler by using what CMake already offers, but I need to understand the exact issue / link failures seen (board / sample / config) so that I know exactly what goes wrong today.

keith-packard commented on cmake/linker/arcmwdt/target.cmake:
```diff
@@ -99,6 +99,7 @@ function(toolchain_ld_link_elf)
     ${LINKERFLAGPREFIX}--whole-archive
     ${ZEPHYR_LIBS_PROPERTY}
     ${LINKERFLAGPREFIX}--no-whole-archive
+    ${ZEPHYR_LINK_LIBS_PROPERTY}
```
Yup. I tried doing just that for quite a while. The trouble is that there's a circular dependency between kernel the common C library and the underlying C library, and there's a dependency between the kernel and various generated headers and objects. Cmake is happy to have circular dependencies between static libraries, but as soon as you introduce the headers and objects into the ring, cmake can't make things work. Every time I tried to use cmake dependencies, I'd get stuck with cmake error messages about a circular dependency with a non static-library in the list.

Additionally, you can't have cmake create a dependency from an external library linked with `-l<foo>` back into a Zephyr library. But that's precisely what exists between `-lstdc++` and the common C library bits. I "solved" that for libc by creating a fake library named `./-lc`, as we have to have actual circular cmake dependencies to get the link command to include them enough times. If application developers need to link to such a library, they would need to perform the same hacks in their own code.

For all other libraries, we just need them to go in front of the 'system API' formed by the union of the kernel, the common C library code and the core C library itself. This seems like the most robust solution I could find; it doesn't include external libraries in the delights of circular dependencies, it doesn't require fake library hacks and it more-or-less reflects what we think of as the Zephyr 'system API' set on top of which other libraries should be built.

keith-packard commented on CMakeLists.txt:
```diff
@@ -865,6 +865,9 @@ foreach(zephyr_lib ${ZEPHYR_LIBS_PROPERTY})
   add_dependencies(${zephyr_lib} zephyr_generated_headers)
 endforeach()
 
+# Convenience variable for linkers
+get_property(ZEPHYR_LINK_LIBS_PROPERTY TARGET zephyr_interface PROPERTY LINK_LIBS)
```
Happy to have you help figure out how to get cmake to work here; I poked at this for several hours without success.

tejlmand replied:
Guess I lost track somewhere, but why is this still needed here ?

Locally I can revert 964a7734b39ab32a9f92ef90ba812472406dc86e and there seems to be no issues, at least when building with Zephyr SDK / picolibc.

keith-packard commented on cmake/linker/ld/target.cmake:
```diff
@@ -124,6 +124,12 @@ function(toolchain_ld_link_elf)
   endif()
 
+  get_property(
+    zephyr_link_libraries
```
Sorry, left over bit from when I didn't have the top level CMakeLists.txt file extracting the property.

cfriedt commented:
> I think this can be made much simpler by using what CMake already offers, but I need to understand the exact issue / link failures seen (board / sample / config) so that I know exactly what goes wrong today.

I hope I can help clarify to some extent. Normally, the system libraries (libc, c++ lib, etc) should not have every symbol included. Only the undefined symbols required to satisfy the linker). And specifically, if the application has already provided a symbol, then the system library version should not be pulled in (so the `-l` options are typically last). Link order matters with gcc, so archives, objects and sources that appear first on the command line will be searched for symbols first. Link order is less less ordered with `clang`, but it still respects the `-l` options that appear last, FWIU).

The nice thing about this is that it allows people to very easily implement their own versions of specific functions. E.g. it is very simple to implement a custom version of `malloc()` with surprisingly little effort (although performance can vary greatly).

So I guess (at least from my perspective) it would be good to ensure that `-Wl,--whole-archive` is not used for the libc, since there are, e.g. lib/libc/common versions of various libc functions that should be pulled in before the libc is linked. With that, only the remaining unresolved / required libc symbols are pulled into the application.

keith-packard commented:
> So I guess (at least from my perspective) it would be good to ensure that `-Wl,--whole-archive` is not used for the libc, since there are, e.g. lib/libc/common versions of various libc functions that should be pulled in before the libc is linked. With that, only the remaining unresolved / required libc symbols are pulled into the application.

Except for the Minimal libc, neither libc nor libstdc++ have ever been included in the --whole-archive set. But, you do raise an interesting point -- even if we don't make the whole change proposed here, we should at least make the minimal C library a regular library using `add_library` -- that will  solve a problem where any data in the minimal C library appears in every application, even if they don't use the related functions.

The problem I'm trying to solve here is for code which glues the C library to the kernel. That's the `libc-hooks.c` stuff for newlib and picolibc, along with the new 'common C library code'. And some of what is going into the POSIX API will also fall into this group. Those libraries implement functions on top of standard Zephyr kernel interfaces _which are then called by  the C library_. We've hacked around this circular reference loop by placing these glue libraries within the `--whole-archive` set and let the linker sort out what is used and what isn't.

Cmake is happy to deal with circular static library references, but it has some strict requirements -- they all have to be "real" cmake library objects created with `add_library`, and nothing but static libraries can be in the loop. Finally, any libraries referenced with `-l` must follow all of the circular reference loop -- those are not `add_library` libraries, so they will never have dependencies which would place them within the loop.

I have a hack that makes the circular references work for libc when referenced with `-lc`. We could formalize this hack and make it work for others as well. However, it seemed simpler to me to change the `toolchain_ld_link_elf` function to insert all of the `zephyr_link_library` bits ahead of the circular reference loop. That works for `-lstdc++` as well as all of the module libraries. That change has an ugly hack in it to split out the libraries from the other link options. We should split the link options out to a separate function so that hack could be removed.

The final struggle I'm having is with the dependency between the `kernel` and `zephyr_generated_headers`. Because the zephyr generated headers aren't a static library, if they ever end up with a dependency on anything in the circular library list, that violates one of the constraints cmake places on circular references and cmake will fail to resolve the dependency graph.

At this point, it feels like there isn't enough interest in solving this to make progress though. Maybe others don't think this is worth fixing?

tejlmand commented:
> At this point, it feels like there isn't enough interest in solving this to make progress though. Maybe others don't think this is worth fixing?

Thanks for the thorough explanation and your findings so far.
I do believe there is high value in what you're trying to achieve, i'm simply trying to avoid an extra level of complexity unless there is no other way to achieve what we need. Such complexities unfortunately tends to cause other problems later.

So I would really like to see more progress on this 👍 .

tejlmand commented:
> The problem I'm trying to solve here is for code which glues the C library to the kernel. That's the libc-hooks.c stuff for newlib and picolibc, along with the new 'common C library code'.  ....

I've been looking a bit at our CMake code and noticed a couple of sub-optimal way of dealing with standard libs.

But even addressing that doesn't solve circular deps, but it help addressing that specific issue.

It's clear to me that there are circular deps between picolibc glue code, libc hooks, kernel.
And as you do know, then we can use the `LINK_INTERFACE_MULTIPLICITY` to increase number of cycles.

It's also clear that linking to the standard libs may cause additional symbols which needs resolving, but it's less clear how many loops we might need for the standard libs (`-lc -lgcc -lstdc++`).
If we just need a single, such as `kernel -c <glue / hook libs> kernel <glue / hook libs> ... -lc`, then things are relatively easy, but it depends on whether it's kernel lib which triggers circular deps to `-lc` (and friends) or if it is the picolibc glue which triggers / requires additional symbols to be resolved.
Or a we in a case where `-lc` (and friends) are needed in each loop ?

keith-packard commented:
Ok, with some help from @tejlmand , I've cleaned up this series a bit.

 * Got rid of the hacks related to `ZEPHYR_LINK_LIBS_PROPERTY`. Now this contains all of the values given to `zephyr_link_libraries` -- libraries, object files and linker options. What this results in is moving all of these bits ahead of 'kernel' in the link command instead of letting them flow to the end. This allows any of these libraries to depend upon the C library API.
 * Eliminate the `PicolibcBuild` hacks that pulled building the picolibc module ahead of the rest of the system. Now that the C library is managed as a 'regular' dependency, this happens naturally as the rest of the system depends up on it.
 * Eliminate `zephyr_libc_link_libraries`. We don't need this any longer; it was only used by picolibc to push -lc and -lgcc past the kernel in the link command line without using normal cmake mechanisms.

The two remaining kludges are:

 1. `zephyr_link_libraries` values are manually inserted into the link command in `toolchain_ld_link_elf`. We need to do this because we cannot ensure that everything in this list is an internal CMake library created with `add_library`, which means we can't create dependencies from those to the C library API.
 2. Toolchain libraries, such as `-lc` and `-lgcc` are turned into STATIC libraries using `add_library` with a hack that creates files with their linker command line value so that we can create dependencies between those and the other parts of the C library dependency graph.

Given the cleanups above, the pretty significant memory savings generated by not including these bits under the `--whole-archive` set, I think it's a net win. I wish the kludges could be eliminated, but I think I've explored all of the available options for doing so and came up empty handed.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -117,6 +117,10 @@ function(zephyr_libc_link_libraries)
   set_property(TARGET zephyr_interface APPEND PROPERTY LIBC_LINK_LIBRARIES ${ARGV})
 endfunction()
 
+function(zephyr_kernel_link_libraries)
+  set_property(TARGET zephyr_interface APPEND PROPERTY KERNEL_LINK_LIBRARIES ${ARGV})
+endfunction()
```
The more I think about this, the more I have second thoughts.
Here are some of my concerns:
- Using the zephyr_interface to pass information to the kernel lib seems a bit wrong, and only reason I can see for the is because `zephyr_interface` happens to be defined early, whereas `kernel` target is not.
- Using a custom property prevents the use of generator expressions, which could be useful in some cases
- Exposing `zephyr_kernel_link_libraries` to the public bears an additional risk of misuse. We both know the distinction between zephyr and kernel, but a lot of less competent users may not, and consider that Zephyr == kernel, and therefore they might wrongly start to use `zephyr_kernel_link_libraries` when they should use `zephyr_link_libraries`.

As the function is only used internally in the `lib/libc` and for C library linking, then I would prefer a `target_link_library(kernel <c-lib>)`.
Something which would also mean correct transfer of includes, compile options, etc. based on public / interface settings of the lib being linked against. In short, regular CMake behavior.

I acknowledge such a request requires that kernel lib is defined earlier in the CMake process, but in fact we need to do something related to kernel lib anyway, which will be described in other comment.

keith-packard replied:
That sounds pretty simple then ; probably just have the top-level CMakeLists.txt file define `kernel` early in the process. It's not like we won't always need that one :-)

keith-packard replied:
That took a bit of a kludge -- we need to know at the time of `add_library` whether we're adding a real library or just an INTERFACE. I moved the logic for figuring that out into the top-level CMakeLists.txt and carried that data along in a property on the kernel target for use in kernel/CMakeLists.txt. Otherwise, this seems like a reasonable cleanup.

tejlmand reviewed (changes requested):
Good progress.

One observation is that we need to handle the pre-built Zephyr kernel library feature here:
https://github.com/zephyrproject-rtos/zephyr/blob/7228642c724a6164a7ce1cdd7a82c3ccc1eb412c/kernel/CMakeLists.txt#L36-L43

because when using a pre-built kernel lib, the circular deps are not handled correctly, as can be seen:
```
$ ninja -v
...
... -Wl,--no-whole-archive
  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj
...
  -lgcc  -Wl,--print-memory-usage  zephyr/arch/common/libisr_tables.a
  -lc  zephyr/lib/libc/picolibc/libc_picolibc.a  -lgcc
  -lc  zephyr/lib/libc/picolibc/libc_picolibc.a  -lgcc
  -lc  zephyr/lib/libc/picolibc/libc_picolibc.a
  zephyr/lib/libc/common/libc_common.a
  /projects/github/ncs/zephyr/kernel/libkernel.a  -mcpu=cortex-m4 ...
```

So if anyway have to touch this area, then probalby do the code properly and take into consideration my other kernel lib comment.

keith-packard committed:
cmake: Use target_link_libraries directly for zephyr_libc_link_libraries

With zephyr_link_libraries now redirected to push libraries before the
kernel, we need to provide a mechanism to push libraries *after* the kernel
so that libc will link correctly. Do that by using what
zephyr_link_libraries used to do, which is to call target_link_libraries.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/newlib: Use target_libc_link_libraries

Switch from zephyr_link_libraries (which is placed before the kernel) to
zephyr_libc_link_libraries (which is placed after).

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
I had to add `zephyr_libc_link_libraries` back in to support newlib. We could convert that to work like picolibc at some point, which would let us remove it at last.

I also had to add a kludge to scan the values passed to zephyr_link_libraries for targets -- we need to extract any of those and make sure they get passed to `target_link_libraries` so that the include paths get added. Maybe there's a mechanism which adds just the include paths but doesn't also insert them into the link command?

tejlmand commented:
> > get_property(ZEPHYR_LINK_LIBS_PROPERTY TARGET zephyr_interface PROPERTY LINK_LIBS)
> 
> This is how we pass any libraries given to `zephyr_link_libraries` to the linker. It's used in `toolchain_ld_link_elf` to place those parameters in front of `kernel`.

yes, I can see what it does. But why are we having a problem in this case ?
Do you have a sample which triggers a wrong linking.

The reason why i'm a bit sensitive on this is because switching to use `get_property()` actually break a fundamental part of CMake because any calls to `zephyr_link_libraries()` after this line:
https://github.com/zephyrproject-rtos/zephyr/blob/b8a1eb8d8e08ab14a50314ac6736d3db45799ab8/CMakeLists.txt#L878

will not be picked up and used for linking.

For example it means that code like this:
```
find_package(Zephyr)

zephyr_link_libraries(<extra_link_flags>)
```

will no longer work as expected.
This can easily be verified with adjusting any sample like this:
```
find_package(Zephyr)

zephyr_link_libraries(break_linking)
```

On main, linking breaks as expected, indicating the link flag was correctly picked up:
```
$ ninja
...
/opt/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: cannot find -lbreak_linking: No such file or directory
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

with this PR, and same line added, line is ignored so linking succeeds (which it shouldn't have in this demo example):
```
$ ninja
[20/21] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       19948 B         1 MB      1.90%
...
```

So if I understand the need correctly, then we (of course I would say) must ensure that `kernel` (and thus c libs) are actually placed last in the link invocation, much like what `CMAKE_C_STANDARD_LIBRARIES` does, but we would need it to handle kernel and all circular dependencies in addition.

Only thing which puzzles me is why is this issue popping up now with the C lib changes.
Is it because moving c glue / hooks out of whole-archive actually increases the circular reference issue, and thereby we are hit harder by the fact that std c libs are only linked once on main:
https://github.com/zephyrproject-rtos/zephyr/blob/c1dfa9796080de9066925736ff6d1fef0433f732/cmake/linker/ld/target.cmake#L143

keith-packard commented:
> > > get_property(ZEPHYR_LINK_LIBS_PROPERTY TARGET zephyr_interface PROPERTY LINK_LIBS)
> > 
> > 
> > This is how we pass any libraries given to `zephyr_link_libraries` to the linker. It's used in `toolchain_ld_link_elf` to place those parameters in front of `kernel`.
> 
> yes, I can see what it does. But why are we having a problem in this case ? Do you have a sample which triggers a wrong linking.

No, we've "fixed" all of the problems by stuffing all of the Zephyr-specific C library bits inside the `--whole-archive` list. The thing which triggered this was discovering that because the C common library is in that list, we _always_ get the malloc initialization code included in the executable, even if there are no calls to malloc in the application. There are numerous additional bits of code and data from the common C library which get included solely because they are seen by the linker due to `--whole-archive` and then get 'stuck' due to some KEEP rule or other reason.
> For example it means that code like this:
> 
> ```
> find_package(Zephyr)
> 
> zephyr_link_libraries(<extra_link_flags>)
> ```
> 
> will no longer work as expected.

But this example *is already broken*. Without that library having dependencies on the kernel, cmake will float it to the end of the linker command line and it will fail to find any kernel (and possibly C library) APIs that it uses.

 I've fiddled with the C library build rules numerous times to make sure that the C library appears after all module libraries, but it's not robust as all of those hacks rely on either having the needed APIs used by some other part of the system, or using `zephyr_libc_link_libraries` which tries hard to get the C library and libgcc appended after the rest of the system by dint of calling `zephyr_link_libraries` late in the top-level CMakeLists.txt -- and that won't work with this example.

> So if I understand the need correctly, then we (of course I would say) must ensure that `kernel` (and thus c libs) are actually placed last in the link invocation, much like what `CMAKE_C_STANDARD_LIBRARIES` does, but we would need it to handle kernel and all circular dependencies in addition.

Yes, exactly. The only mechanism I can see for emulating `CMAKE_C_STANDARD_LIBRARIES` is to manually insert all of the 'other' libraries ahead of the circular list of Zephyr system API libraries.

> Only thing which puzzles me is why is this issue popping up now with the C lib changes. Is it because moving c glue / hooks out of whole-archive actually increases the circular reference issue, and thereby we are hit harder by the fact that std c libs are only linked once on main:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/c1dfa9796080de9066925736ff6d1fef0433f732/cmake/linker/ld/target.cmake#L143

That's only part of it -- the other piece is that any module or drivers using `zephyr_link_libraries` will not get access to anything from the common C library or C library stubs unless some other part of the system has accidentally used code from the same file.

As a simple example, I want this to end up using 'malloc' from the C common library, but without requiring that library be under the whole-archive umbrella:

```
#include <stdio.h>

int main(void)
{
    char *s;
    asprintf(&s, "hello world %s\n", CONFIG_BOARD);
    printf("%s", s);
    return 0;
}
```

`asprintf` is a C library API which uses `malloc`, which is provided by the common C library and uses `sys_heap_aligned_alloc`, which is provided by the kernel.

Note that this example is "broken" in not calling `free`. However, if it did call `free`, then that would pull in `malloc.c` from the common C library, making `malloc` available before the linker ever tried to resolve it when relocating `asprintf`.

tejlmand commented on CMakeLists.txt:
```diff
@@ -544,6 +544,22 @@ if(EXISTS ${APPLICATION_SOURCE_DIR}/VERSION)
+
+set(libkernel ${CMAKE_CURRENT_SOURCE_DIR}/kernel/libkernel${CMAKE_STATIC_LIBRARY_SUFFIX})
+if(EXISTS ${libkernel})
+  add_library(kernel INTERFACE)
+  set_property(TARGET kernel PROPERTY KERNEL_INTERFACE TRUE)
```
thanks for this.

Let's use the opportunity to import pre-built libs the proper way:
```suggestion
  add_library(kernel STATIC IMPORTED GLOBAL)
  set_property(TARGET kernel PROPERTY IMPORTED_LOCATION ${libkernel})
```

tejlmand commented on CMakeLists.txt:
```diff
@@ -544,6 +544,22 @@ if(EXISTS ${APPLICATION_SOURCE_DIR}/VERSION)
+  set_property(TARGET kernel PROPERTY KERNEL_INTERFACE TRUE)
+else()
+  add_library(kernel)
+  set_property(TARGET kernel PROPERTY KERNEL_INTERFACE FALSE)
```
Line not needed, CMake already has various properties to to get such information from a target.
```suggestion
```

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -30,16 +30,10 @@ zephyr_syscall_header_ifdef(
-zephyr_library_get_current_dir_lib_name(${ZEPHYR_BASE} libkernel_stem)
-set(libkernel ${CMAKE_CURRENT_SOURCE_DIR}/lib${libkernel_stem}${CMAKE_STATIC_LIBRARY_SUFFIX})
-unset(libkernel_stem)
+get_target_property(KERNEL_INTERFACE kernel KERNEL_INTERFACE)
```
now that we import the kernel lib the proper way, then let's use the CMake built-in IMPORTED property instead.
```suggestion
get_target_property(is_imported kernel IMPORTED)
```

keith-packard replied:
Nice cleanup. I've pushed this update.

tejlmand reviewed (dismissed):
Review of the kernel library cleanup.

Nice cleanup 👍

tejlmand commented:
> But this example is already broken. Without that library having dependencies on the kernel, cmake will float it to the end of the linker command line and it will fail to find any kernel (and possibly C library) APIs that it uses.

yes, but even the proposal to use properties will not fix that (only make other parts worse, like not even picking up additional linker flags).
Regarding linking libraries, then I also mention that here: https://github.com/zephyrproject-rtos/zephyr/pull/51949#discussion_r1168393141

For example, any:
```
find_package(Zephyr)

add_library(foo <files>)
target_link_libraries(app foo)
```
can result in lib `foo` to float beyond `kernel`, which is why a lib which depends on kernel functions must also do:
```
target_link_libraries(foo kernel)
```

Now, one can say that linking against kernel is acceptable if you're actually using functions from the kernel, but it shouldn't be necessary to link against std C libs (or kernel to get std C libs linked later)

So we still need to work on an improved solution.

keith-packard commented:
> yes, but even the proposal to use properties will not fix that (only make other parts worse, like not even picking up additional linker flags). 

Right, using the property makes it work for anything managed within the Zephyr cmake world. So it's better for some people at least?

> Now, one can say that linking against kernel is acceptable if you're actually using functions from the kernel, but it shouldn't be necessary to link against std C libs (or kernel to get std C libs linked later)

Hrm. Maybe when we've got `zephyr_link_options` split out we could walk the parameters to `zephyr_link_libraries` and automatically insert a dependency on the C library. That just requires always having a real C library, but we're almost there now. That makes `zephyr_link_libraries` look like this:
```
function(zephyr_link_libraries)
  target_link_libraries(zephyr_interface INTERFACE ${ARGV})
  foreach(zephyr_link_lib ${ARGV})
    if(TARGET ${zephyr_link_lib})
      target_link_libraries(${zephyr_link_lib} PRIVATE c)
    endif()
  endforeach()
endfunction()
```

keith-packard commented:
Still doesn't solve people using `target_link_library` directly, but maybe this is sufficient? At least it doesn't depend upon calling this before the Zephyr CMakeLists.txt extracts the property ...

keith-packard commented:
And runs smack into circular dependency problems with `zephyr_generated_headers`. 

https://github.com/keith-packard/zephyr/tree/circular-link-libraries

has the proposed code on the latest commit, along with the instructions to reproduce the failure. This definitely looks like something only a cmake expert can solve...

keith-packard commented:
whoa, passing all of the tests at last. I'm happy to keep poking at this, but am wondering if this is a 'good enough' improvement to get merged while we work on finding a way to get regular Cmake mechanisms for managing external libraries working better.

andyross reviewed (dismissed):
My vote would be to merge this.  The use of --whole-archive is in almost all cases a hack we want to eliminate anyway, as it makes reasoning about linker behavior obscure and complicated (especially given all the platform-dependence of the way we link).

Situations like e.g. SYS_INIT() or IRQ_CONNECT() that rely on it for reference detection should probably just be migrated to kconfig or devicetree as appropriate, where the semantics are well-defined and inspectable.  But that requires figuring out all the edges where we're relying on it, which is fragile...

keith-packard commented:
Rebased and patched to avoid conflicts. Otherwise, no changes.

fabiobaltieri commented:
@tejlmand can you do another pass? We are close to RC1, should wrap this now if we want this in before feature freeze.

keith-packard committed:
libc/common: Add toolchain.h to gmtime_r for ZRESTRICT

When built as a regular library, this file doesn't get included
automatically.

Signed-off-by: Keith Packard <keithp@keithp.com>

cfriedt commented:
I'm still ready to approve this 😁

keith-packard commented:
> I'm still ready to approve this 😁

Thanks; I just rebased today. We'll see if it actually works again...

keith-packard commented:
ok, looks like this still works; the rebase was pretty straightforward, but alas not entirely automatic. Fortunately, the big change to all of the per-linker cmake bits was no longer necessary as there's now a variable carrying the non-whole-archive libs down from the top level. I could split this series into some smaller pieces if that would make it easier to review; a big part of the change here was to break the libc-hooks.c file apart so only the necessary bits would get pulled in. That could be applied independently of the rest of the series.

cfriedt reviewed (dismissed):
Looks reasonable to me

keith-packard commented:
Just resolving a minor merge conflict

cfriedt commented:
@tejlmand - can you please re-review?

keith-packard commented:
I've rebased this series again; would sure be nice to get it merged so we could stop needing per-application configuration for malloc usage.

cfriedt reviewed (dismissed):
Still looks good to me.

It would be nice if there was no need to provide workarounds for circular dependencies, but that's outside the scope of the libc.

A while ago, I threw together something that would only lazy-initialize the minimal libc malloc (i.e. initialize on first use) but there was some issue that wasn't immediately clear to me how to fix.

keith-packard committed:
newlib: Provide _unlink when CONFIG_FILE_SYSTEM is set

_unlink is used by newlib's 'remove' implementation, which is
part of the standard C API, not a POSIX extension.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
newlib: provide _unlink_r for xtensa

This is required for newlib's remove implementation.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
intel_adsp: Skip adding zephyr_generated_headers dependency

I added this kludge to prevent a dependency loop from forming. I have no
idea why this particular library is causing the problem.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
I had to add a kludge to get intel_adsp builds working; that's gonna need someone with cmake knowledge to help resolve, I think.
